tests/conftest.py

```diff
@@ -1066,7 +1066,26 @@ def _ha_orm_quote(mixed, ident):
         assert not sqlalchemy_utils.database_exists(db_url)
         sqlalchemy_utils.create_database(db_url, encoding="utf8")
     yield db_url
-    if db_url.startswith("mysql://") or db_url.startswith("postgresql://"):
+    if db_url.startswith("mysql://"):
+        import sqlalchemy as sa
+
+        made_url = sa.make_url(db_url)
+        db = made_url.database
+        engine = sa.create_engine(db_url)
+        # Kill any open connections to the database before dropping it
+        # to ensure that InnoDB does not deadlock.
+        with engine.begin() as connection:
+            query = sa.text(
+                "select id FROM information_schema.processlist WHERE db=:db and id != CONNECTION_ID()"
+            )
+            for row in connection.execute(query, parameters={"db": db}).fetchall():
+                _LOGGER.warning(
+                    "Killing MySQL connection to temporary database %s", row.id
```

Comment:
Should we instead fail the test here?

I think recorder tests don't clean up after themselves after the merge of https://github.com/home-assistant/core/pull/84681

After that was merged, there are stray prints from SQLAlchemy:

```
pytest -vv tests/components/recorder

...

 tests/components/recorder/test_init.py::test_connect_args_priority[mysql://user:password@SERVER_IP/DB_NAME?charset=utf8mb4] ✓                                                                          31% ███▏
 tests/components/recorder/test_init.py::test_connect_args_priority[mysql://user:password@SERVER_IP/DB_NAME?blah=bleh&charset=other] ✓                                                                  31% ███▏      INFO:sqlalchemy.engine.Engine:INSERT INTO event_data (hash, shared_data) VALUES (?, ?)
INFO:sqlalchemy.engine.Engine:[cached since 0.002658s ago] (3478486156, '{"domain":"recorder","service":"disable"}')
INFO:sqlalchemy.engine.Engine:INSERT INTO event_data (hash, shared_data) VALUES (?, ?)
INFO:sqlalchemy.engine.Engine:[cached since 0.002985s ago] (2705000516, '{"component":"recorder"}')

INFO:sqlalchemy.engine.Engine:INSERT INTO state_attributes (hash, shared_attrs) VALUES (?, ?)
 tests/components/recorder/test_migrate.py::test_schema_update_calls ✓
```

Reply:
Without it it seems tests can randomly cause the CI to deadlock forever.   I think we should merge this fix now, and try to figure out the root cause when its not blocking the CI.

Reply:
Failing here is a good idea if we can solve the cleanup without leaving the CI broken for too long.

I've got to get some more sleep (only woke up to take out the dog .. and that was a few hours ago) so I'm back off to  bed